pkg/kv/cosmosdb/store.go

```diff
@@ -0,0 +1,428 @@
+	"github.com/Azure/azure-sdk-for-go/sdk/data/azcosmos"
+	"github.com/treeverse/lakefs/pkg/ident"
+	"github.com/treeverse/lakefs/pkg/kv"
+	kvparams "github.com/treeverse/lakefs/pkg/kv/params"
+)
+
+type Driver struct{}
+
+type Store struct {
+	containerClient  *azcosmos.ContainerClient
+	consistencyLevel azcosmos.ConsistencyLevel
+}
+
+const (
+	DriverName = "cosmosdb"
+)
+
+//nolint:gochecknoinits
+func init() {
+	kv.Register(DriverName, &Driver{})
+}
+
+// Open - opens and returns a KV store over CosmosDB. This function creates the DB session
+// and sets up the KV table.
+func (d *Driver) Open(ctx context.Context, kvParams kvparams.Config) (kv.Store, error) {
+	params := kvParams.CosmosDB
+	if params == nil {
+		return nil, fmt.Errorf("missing %s settings: %w", DriverName, kv.ErrDriverConfiguration)
+	}
+	if params.Endpoint == "" {
+		return nil, fmt.Errorf("missing endpoint: %w", kv.ErrDriverConfiguration)
+	}
+	if params.Database == "" {
+		return nil, fmt.Errorf("missing database: %w", kv.ErrDriverConfiguration)
+	}
+	if params.Container == "" {
+		return nil, fmt.Errorf("missing container: %w", kv.ErrDriverConfiguration)
+	}
+
+	log.Printf("CosmosDB: connecting to %s", params.Endpoint)
+
+	var client *azcosmos.Client
+	if params.ReadWriteKey != "" {
+		cred, err := azcosmos.NewKeyCredential(params.ReadWriteKey)
+		if err != nil {
+			return nil, fmt.Errorf("creating key: %w", err)
+		}
+
+		// hook for using emulator for testing
+		if params.Client == nil {
+			params.Client = http.DefaultClient
+		}
+		// Create a CosmosDB client
+		client, err = azcosmos.NewClientWithKey(params.Endpoint, cred, &azcosmos.ClientOptions{
+			ClientOptions: azcore.ClientOptions{
+				Transport: params.Client,
+			},
+		})
+		if err != nil {
+			return nil, fmt.Errorf("creating client using access key: %w", err)
+		}
+	} else {
+		cred, err := azidentity.NewDefaultAzureCredential(nil)
+		if err != nil {
+			return nil, fmt.Errorf("default creds: %w", err)
+		}
+		client, err = azcosmos.NewClient(params.Endpoint, cred, nil)
+		if err != nil {
+			return nil, fmt.Errorf("creating client with default creds: %w", err)
+		}
+	}
+
+	dbClient, err := getOrCreateDatabase(ctx, client, params)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create container client
+	containerClient, err := getOrCreateContainer(ctx, dbClient, params)
+	if err != nil {
+		return nil, err
+	}
+
+	cLevel := azcosmos.ConsistencyLevelBoundedStaleness
+	if !params.StrongConsistency {
+		cLevel = azcosmos.ConsistencyLevelSession
+	}
+	return &Store{
+		containerClient:  containerClient,
+		consistencyLevel: cLevel,
+	}, nil
+}
+
+func getOrCreateDatabase(ctx context.Context, client *azcosmos.Client, params *kvparams.CosmosDB) (*azcosmos.DatabaseClient, error) {
+	dbResp, err := client.CreateDatabase(ctx, azcosmos.DatabaseProperties{ID: params.Database}, nil)
+	if err != nil {
+		errCode := errStatusCode(err)
+		if errCode != http.StatusConflict && errCode != http.StatusCreated {
```

Comment:
Question - we are getting `http.StatusCreated` inside `err != nil` when we `client.CreateDatabase` successful?
Trying to see if we need to check for http.StatusCreated?

Reply:
You're right, we don't.